File: answers/leetcode/Scramble String.cpp

class Solution {
public:
	bool isScramble(string s1, string s2) {
		vector<vector<vector<bool> > > dp(s1.size(), vector<vector<bool> > (s1.size(), vector<bool> (s1.size(), false)));
		for (size_t i = 0; i < s1.size() && i < s2.size(); ++ i) {
			for (size_t j = 0; i + j < s1.size() && i + j < s2.size(); ++ j) {
				for (size_t k = 0; i + k < s1.size() && i + k < s2.size(); ++ k) {
					if (s1.substr(j, i + 1) == s2.substr(k, i + 1)) {
						dp[i][j][k] = true;
						continue;
					}
					for (size_t m = 0; m < i; ++ m) {
						if ((dp[m][j][k] && dp[i - m - 1][j + m + 1][k + m + 1]) || (dp[m][j][k + i - m] && dp[i - m - 1][j + m + 1][k])) {
							dp[i][j][k] = true;
							break;
						}
					}
				}
			}
		}
		return dp[s1.size() - 1][0][0];
	}
};

File: answers/leetcode/Scramble String/Scramble String.java
class Solution {
    public boolean isScramble(String s1, String s2) {
        int length = s1.length();
        boolean[][][] scramble = new boolean[length][length][length];
        for (int i = 0; i < length; i++) {
            for (int j = 0; i + j < length; j++) {
                for (int k = 0; i + k < length; k++) {
                    if (i == 0) {
                        scramble[i][j][k] = s1.charAt(j) == s2.charAt(k);
                    } else {
                        for (int m = 0; m < i; m++) {
                            if ((scramble[m][j][k] && scramble[i - m - 1][j + m + 1][k + m + 1])
                                || (scramble[m][j][k + i - m] && scramble[i - m - 1][j + m + 1][k])) {
                                scramble[i][j][k] = true;
                                break;
                            }
                        }
                    }
                    //System.out.println(i + " " + j + " " + k + ": " + s1.substring(j, j + i + 1) + " " + s2.substring(k, k + i + 1) + " " + scramble[i][j][k]);
                }
            }
        }
        return scramble[length - 1][0][0];
    }
}